zebra-consensus/src/primitives/halo2.rs

```diff
@@ -0,0 +1,281 @@
+    /// fallback logic when batch verification fails.
+    pub fn verify_single(&self, vk: &VerifyingKey) -> Result<(), halo2::plonk::Error> {
+        self.proof.verify(vk, &self.instances[..])
+    }
+}
+
+#[derive(Default)]
+pub struct BatchVerifier {
+    queue: Vec<Item>,
+}
+
+impl BatchVerifier {
+    pub fn queue(&mut self, item: Item) {
+        self.queue.push(item);
+    }
+
+    pub fn verify<R: RngCore + CryptoRng>(
+        self,
+        _rng: R,
+        vk: &VerifyingKey,
+    ) -> Result<(), halo2::plonk::Error> {
+        for item in self.queue {
+            item.verify_single(vk)?;
+        }
+
+        Ok(())
+    }
+}
+
+// === END TEMPORARY BATCH HALO2 SUBSTITUTE ===
+
+impl From<zebra_chain::orchard::ShieldedData> for Item {
+    fn from(shielded_data: zebra_chain::orchard::ShieldedData) -> Item {
+        use orchard::{circuit, note, primitives::redpallas, tree, value};
+
+        let anchor = tree::Anchor::from_bytes(shielded_data.shared_anchor.into()).unwrap();
+
+        let enable_spend = shielded_data
+            .flags
+            .contains(zebra_chain::orchard::Flags::ENABLE_SPENDS);
+        let enable_output = shielded_data
+            .flags
+            .contains(zebra_chain::orchard::Flags::ENABLE_OUTPUTS);
+
+        let instances = shielded_data
+            .actions()
+            .map(|action| {
+                circuit::Instance::from_parts(
+                    anchor,
+                    value::ValueCommitment::from_bytes(&action.cv.into()).unwrap(),
+                    note::Nullifier::from_bytes(&action.nullifier.into()).unwrap(),
+                    redpallas::VerificationKey::<redpallas::SpendAuth>::try_from(<[u8; 32]>::from(
+                        action.rk,
+                    ))
+                    .unwrap(),
+                    note::ExtractedNoteCommitment::from_bytes(&action.cm_x.into()).unwrap(),
+                    enable_spend,
+                    enable_output,
+                )
+            })
+            .collect();
+
+        Item {
+            instances,
+            proof: orchard::circuit::Proof::new(shielded_data.proof.0),
+        }
+    }
+}
+
+/// An error that may occur when verifying [Halo2 proofs of Zcash Orchard Action
+/// descriptions][actions].
+///
+/// [actions]: https://zips.z.cash/protocol/protocol.pdf#actiondesc
+// TODO: if halo2::plonk::Error gets the std::error::Error trait derived on it,
+// remove this and just wrap `halo2::plonk::Error` as an enum variant of
+// `crate::transaction::Error`, which does the trait derivation via `thiserror`
+#[derive(Clone, Debug, Error, Eq, PartialEq)]
+pub enum Halo2Error {
+    /// The constraint system is not satisfied.
+    // #[default]
+    ConstraintSystemFailure,
+    /// Catchall for now until https://github.com/zcash/halo2/pull/394 is merged
+    Other,
+}
+
+impl fmt::Display for Halo2Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(f, "{}", self)
```

Comment:
Isn't this an infinite loop?

Reply:
You're right!

![image](https://user-images.githubusercontent.com/552961/141388020-0b8189c6-0842-4a35-b1d5-d0a9176c7425.png)

![image](https://user-images.githubusercontent.com/552961/141388050-ea3b5467-8549-48c6-9ad2-f6bbc43843d2.png)

Reply:
Fixed in 16b2c2b2